pkg/apis/eventing/v1alpha1/source_types.go

```diff
@@ -0,0 +1,171 @@
+
+	// Status is the current status of the Source.
+	// +optional
+	Status SourceStatus `json:"status,omitempty"`
+}
+
+// Check that Source can be validated and can be defaulted.
+var _ webhook.GenericCRD = (*Source)(nil)
+
+// Check that SourceStatus may have its conditions managed.
+var _ duckv1alpha1.ConditionsAccessor = (*SourceStatus)(nil)
+
+// Check that Source implements the Conditions duck type.
+var _ = duck.VerifyType(&Source{}, &duckv1alpha1.Conditions{})
+
+// Check that Source implements the Generation duck type.
+var emptyGenSource duckv1alpha1.Generation
+var _ = duck.VerifyType(&Source{}, &emptyGenSource)
+
+// And it's Subscribable
+var _ = duck.VerifyType(&Subscription{}, &duckv1alpha1.Subscribable{})
+
+// SourceSpec is the spec for a Source resource.
+type SourceSpec struct {
+	// TODO: Generation does not work correctly with CRD. They are scrubbed
+	// by the APIserver (https://github.com/kubernetes/kubernetes/issues/58778)
+	// So, we add Generation here. Once that gets fixed, remove this and use
+	// ObjectMeta.Generation instead.
+	// +optional
+	Generation int64 `json:"generation,omitempty"`
+
+	// Provisioner is used to create any backing resources and configuration.
+	// +required
+	Provisioner *ProvisionerReference `json:"provisioner,omitempty"`
+
+	// Arguments defines the arguments to pass to the Provisioner which provisions
+	// this Source.
+	// +optional
+	Arguments *runtime.RawExtension `json:"arguments,omitempty"`
+
+	// Specify an existing channel to use to emit events. If empty, create a new
+	// Channel using the cluster/namespace default.
+	//
+	// This object must fulfill the Channelable contract.
+	//
+	// You can specify only the following fields of the ObjectReference:
+	//   - Kind
+	//   - APIVersion
+	//   - Name
+	// Currently Kind must be "Channel" and
+	// APIVersion must be "eventing.knative.dev/v1alpha1"
+	// +optional
+	Channel *corev1.ObjectReference `json:"target,omitempty"`
+}
+
+const (
+	// SourceConditionReady has status True when the Source is ready to send events.
+	SourceConditionReady = duckv1alpha1.ConditionReady
+
+	// SourceConditionProvisioned has status True when the Source's backing
+	// resources have been provisioned.
+	SourceConditionProvisioned duckv1alpha1.ConditionType = "Provisioned"
+)
+
+var sourceCondSet = duckv1alpha1.NewLivingConditionSet(SourceConditionProvisioned)
+
+// SourceStatus is the status for a Source resource
+type SourceStatus struct {
+	// Conditions holds the state of a source at a point in time.
+	// +optional
+	// +patchMergeKey=type
+	// +patchStrategy=merge
+	Conditions duckv1alpha1.Conditions `json:"conditions,omitempty" patchStrategy:"merge" patchMergeKey:"type"`
+
+	// ObservedGeneration is the 'Generation' of the Source that
+	// was last reconciled by the controller.
+	// +optional
+	ObservedGeneration int64 `json:"observedGeneration,omitempty"`
+
+	// Source might be Subscribable. This points to the Channelable object.
```

Comment:
might? I think it must be Subscribable, how else would we get events from it? Is this CRD the "factory" for say, creating K8S events for consumption by the system, or is an instance one that I could create a Subscription.Spec.From to point to?